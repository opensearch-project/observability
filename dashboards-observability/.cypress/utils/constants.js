/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const delay = 1500;
export const COMMAND_TIMEOUT_LONG = 10000;

// trace analytics
export const TRACE_ID = '8832ed6abbb2a83516461960c89af49d';
export const SPAN_ID = 'a673bc074b438374';
export const SERVICE_NAME = 'frontend-client';
export const SERVICE_SPAN_ID = '7df5609a6d104736';

export const testDataSet = [
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-service-map-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-service-map.json',
    index: 'otel-v1-apm-service-map',
  },
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001.json',
    index: 'otel-v1-apm-span-000001',
  },
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000002.json',
    index: 'otel-v1-apm-span-000002',
  },
]

export const setTimeFilter = (setEndTime = false, refresh = true) => {
  const startTime = 'Mar 25, 2021 @ 10:00:00.000';
  const endTime = 'Mar 25, 2021 @ 11:00:00.000';
  cy.get('button.euiButtonEmpty[aria-label="Date quick select"]').click();
  cy.get('.euiQuickSelect__applyButton').click();
  cy.get('.euiSuperDatePicker__prettyFormatLink').click();
  cy.get(
    'button.euiDatePopoverButton--start[data-test-subj="superDatePickerstartDatePopoverButton"]'
  ).click();
  cy.get('.euiTab__content').contains('Absolute').click();
  cy.get('input[data-test-subj="superDatePickerAbsoluteDateInput"]')
    .focus()
    .type('{selectall}' + startTime, { force: true });
  if (setEndTime) {
    cy.wait(delay);
    cy.get(
      'button.euiDatePopoverButton--end[data-test-subj="superDatePickerendDatePopoverButton"]'
    ).click();
    cy.wait(delay);
    cy.get('.euiTab__content').contains('Absolute').click();
    cy.get('input[data-test-subj="superDatePickerAbsoluteDateInput"]')
      .focus()
      .type('{selectall}' + endTime, { force: true });
  }
  if (refresh) cy.get('.euiButton__text').contains('Refresh').click();
  cy.wait(delay);
};

// notebooks
export const TEST_NOTEBOOK = 'Test Notebook';
export const SAMPLE_URL = 'https://github.com/opensearch-project/sql/tree/main/sql-jdbc';
export const NOTEBOOK_TEXT = 'Use Notebooks to interactively and collaboratively develop rich reports backed by live data. Common use cases for notebooks includes creating postmortem reports, designing run books, building live infrastructure reports, or even documentation.';
export const OPENSEARCH_URL = 'https://opensearch.org/docs/latest/observability-plugin/notebooks/';
export const MARKDOWN_TEXT = `%md
# Heading 1

#### List and links

* 1
* 2
* [SQL JDBC](${SAMPLE_URL})

---
#### Code block
* Explain SQL
\`\`\`
POST _plugins/_sql/_explain
{
  "query": "SELECT * FROM my-index LIMIT 50"
}
\`\`\`

#### Table
| a1 | b1 | c1 | d1 |
|----|----|----|----|
| a2 | b2 | c2 | d2 |
| a3 | b3 | c3 | d3 |
`

export const SQL_QUERY_TEXT = `%sql
select * from opensearch_dashboards_sample_data_flights limit 20
`

export const PPL_QUERY_TEXT = `%ppl
source=opensearch_dashboards_sample_data_flights
`

export const supressResizeObserverIssue = () => {
  // exception is thrown on loading EuiDataGrid in cypress only, ignore for now
  cy.on('uncaught:exception', (err, runnable) => {
    if (err.message.includes('ResizeObserver loop')) return false;
  });
};

export const verify_traces_spans_data_grid_cols_exists = () => {
  cy.get('.euiDataGridHeaderCell__content').contains('Span ID').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Trace ID').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Operation').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Duration').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Start time').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('End time').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Errors').should('exist');
}

export const count_table_row = (expected_row_count) => {
  cy.get('.euiDataGridHeader [role="columnheader"]').then($el => {
    let colmun_header_count = Cypress.$($el).length;
    let table_grid_cell_count = Cypress.$('[data-test-subj="dataGridRowCell"]').length;
    const total_row_count = table_grid_cell_count / colmun_header_count;
    expect(total_row_count).to.equal(expected_row_count)
  });
}
