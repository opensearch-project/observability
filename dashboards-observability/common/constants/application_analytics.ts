/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const TAB_OVERVIEW_ID = 'app-analytics-overview';
export const TAB_SERVICE_ID = 'app-analytics-service';
export const TAB_TRACE_ID = 'app-analytics-trace';
export const TAB_LOG_ID = 'app-analytics-log';
export const TAB_PANEL_ID = 'app-analytics-panel';
export const TAB_CONFIG_ID = 'app-analytics-config';
export const TAB_OVERVIEW_TITLE = 'Overview';
export const TAB_SERVICE_TITLE = 'Services';
export const TAB_TRACE_TITLE = 'Traces & Spans';
export const TAB_LOG_TITLE = 'Log Events';
export const TAB_PANEL_TITLE = 'Panel';
export const TAB_CONFIG_TITLE = 'Configuration';

export const APP_ANALYTICS_API_PREFIX = '/api/observability/application';
