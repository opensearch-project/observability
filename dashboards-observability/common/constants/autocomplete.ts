/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const firstCommand = [{ label: 'source' }];

export const pipeCommands = [
  { label: 'dedup' },
  { label: 'eval' },
  { label: 'fields' },
  { label: 'head' },
  { label: 'rare' },
  { label: 'rename' },
  { label: 'sort' },
  { label: 'stats' },
  { label: 'top' },
  { label: 'where' },
];

export const statsCommands = [
  { label: 'count()' },
  { label: 'sum(' },
  { label: 'avg(' },
  { label: 'max(' },
  { label: 'min(' },
  { label: 'var_samp(' },
  { label: 'var_pop(' },
  { label: 'stddev_samp(' },
  { label: 'stddev_pop(' },
];

export const numberTypes = [
  'long',
  'integer',
  'short',
  'byte',
  'double',
  'float',
  'half_float',
  'scaled_float',
  'unsigned_long',
];

export interface AutocompleteItem {
  input: string;
  itemName: string;
  label: string;
  suggestion: string;
  __autocomplete_id?: number;
}

export interface FieldItem {
  label: string;
  type: string;
}

export interface IndexItem {
  label: string;
}

export interface DataItem {
  label: string;
  doc_count: any;
}

const JUST_SOURCE_REGEX = new RegExp('source = (\\S+)');
const SOURCE_WHERE_REGEX = new RegExp('source = (\\S+) \\| where \\S+ = \\S+');
const SOURCE_MATCH_REGEX = new RegExp('source = (\\S+) \\| where match\\(\\S+, \\S+\\)');
export const EMPTY_REGEX = new RegExp('\\s+');
export const MATCH_FIELD_AFTER_WHERE = new RegExp('\\s*where\\s*\\S*');
export const EQUAL_AFTER_FIELD = new RegExp('\\s*where\\s*(\\S+)\\s+');
export const DATA_AFTER_EQUAL = new RegExp('\\s*where\\s*\\S+\\s*=\\s*\\S*');
export const PIPE_AFTER_WHERE = new RegExp('\\s*where\\s*\\S+\\s*=\\s*\\S+\\s+');
export const FIELD_AFTER_MATCH = new RegExp('\\s*where\\s*match\\(\\s*\\S*');
export const COMMA_AFTER_FIELD = new RegExp('\\s*where\\s*match\\(\\s*(\\S+)\\s+');
export const DATA_AFTER_COMMA = new RegExp('\\s*where\\s*match\\(\\s*\\S+\\s*,\\s*');
export const CLOSE_AFTER_DATA = new RegExp('\\s*where\\s*match\\(\\s*\\S+\\s*,\\s*\\S+\\s+');
export const PIPE_AFTER_MATCH = new RegExp(
  '\\s*where\\s*match\\(\\s*\\S+\\s*,\\s*\\S+\\s*\\S+\\s*\\)\\s*'
);
export const FIELD_AFTER_DEDUP = new RegExp('\\s*dedup\\s*\\S*');
export const PIPE_AFTER_FIELD = new RegExp('\\s*dedup\\s*\\S+\\s+');

export const regexForSuggestion = [
  PIPE_AFTER_FIELD,
  FIELD_AFTER_DEDUP,
  PIPE_AFTER_MATCH,
  CLOSE_AFTER_DATA,
  DATA_AFTER_COMMA,
  COMMA_AFTER_FIELD,
  FIELD_AFTER_MATCH,
  PIPE_AFTER_WHERE,
  DATA_AFTER_EQUAL,
  EQUAL_AFTER_FIELD,
  MATCH_FIELD_AFTER_WHERE,
  EMPTY_REGEX,
];

export const regexForIndex = [JUST_SOURCE_REGEX, SOURCE_WHERE_REGEX, SOURCE_MATCH_REGEX];
