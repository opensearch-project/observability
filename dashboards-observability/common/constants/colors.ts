/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { colorPalette } from '@elastic/eui';

export const BLUES_PALETTE = {
  name: 'Blues',
  label: 'Blues',
  colors: [
    'rgb(5,10,172)',
    'rgb(40,60,190)',
    'rgb(70,100,245)',
    'rgb(90,120,245)',
    'rgb(106,137,247)',
    'rgb(220,220,220)',
  ],
};

export const REDS_PALETTE = {
  name: 'Reds',
  label: 'Reds',
  colors: ['rgb(220,220,220)', 'rgb(245,195,157)', 'rgb(245,160,105)', 'rgb(178,10,28)'],
};

export const GREENS_PALETTE = {
  name: 'Greens',
  label: 'Greens',
  colors: [
    'rgb(0,68,27)',
    'rgb(0,109,44)',
    'rgb(35,139,69)',
    'rgb(65,171,93)',
    'rgb(116,196,118)',
    'rgb(161,217,155)',
    'rgb(199,233,192)',
    'rgb(229,245,224)',
    'rgb(247,252,245)',
  ],
};

export const GREYS_PALETTE = {
  name: 'Greys',
  label: 'Greys',
  colors: ['rgb(0,0,0)', 'rgb(255,255,255)'],
};

export const BLUE_RED_PALETTE = {
  name: 'Bluered',
  label: 'Blue-Red',
  colors: ['rgb(0,0,255)', 'rgb(255,0,0)'],
};

export const RED_BLUE_PALETTE = {
  name: 'RdBu',
  label: 'Red-Blue',
  colors: [
    'rgb(5,10,172)',
    'rgb(106,137,247)',
    'rgb(190,190,190)',
    'rgb(220,170,132)',
    'rgb(230,145,90)',
    'rgb(178,10,28)',
  ],
};

export const YELLOW_ORANGE_RED_PALETTE = {
  name: 'YlOrRd',
  label: 'Yellow-Orange-Red',
  colors: [
    'rgb(128,0,38)',
    'rgb(189,0,38)',
    'rgb(227,26,28)',
    'rgb(252,78,42)',
    'rgb(253,141,60)',
    'rgb(254,178,76)',
    'rgb(254,217,118)',
    'rgb(255,237,160)',
    'rgb(255,255,204)',
  ],
};

export const YELLOW_GREEN_BLUE_PALETTE = {
  name: 'YlGnBu',
  label: 'Yellow-Green-Blue',
  colors: [
    'rgb(8,29,88)',
    'rgb(37,52,148)',
    'rgb(34,94,168)',
    'rgb(29,145,192)',
    'rgb(65,182,196)',
    'rgb(127,205,187)',
    'rgb(199,233,180)',
    'rgb(237,248,217)',
    'rgb(255,255,217)',
  ],
};

export const DEFAULT_PALETTE = 'default';
export const SINGLE_COLOR_PALETTE = 'singleColor';
export const MULTI_COLOR_PALETTE = 'multicolor';

export const COLOR_PALETTES = [
  {
    value: DEFAULT_PALETTE,
    title: 'Default',
    type: 'text',
  },
  {
    value: SINGLE_COLOR_PALETTE,
    title: 'Single color',
    type: 'text',
  },
  {
    value: MULTI_COLOR_PALETTE,
    title: 'Multicolored',
    type: 'text',
  },
  {
    value: BLUES_PALETTE.name,
    title: BLUES_PALETTE.label,
    palette: colorPalette(BLUES_PALETTE.colors, 20),
    type: 'gradient',
  },
  {
    value: REDS_PALETTE.name,
    title: REDS_PALETTE.label,
    palette: colorPalette(REDS_PALETTE.colors, 20),
    type: 'gradient',
  },
  {
    value: GREENS_PALETTE.name,
    title: GREENS_PALETTE.label,
    palette: colorPalette(GREENS_PALETTE.colors, 20),
    type: 'gradient',
  },
  {
    value: GREYS_PALETTE.name,
    title: GREYS_PALETTE.label,
    palette: colorPalette(GREYS_PALETTE.colors, 20),
    type: 'gradient',
  },
  {
    value: BLUE_RED_PALETTE.name,
    title: BLUE_RED_PALETTE.label,
    palette: colorPalette(BLUE_RED_PALETTE.colors, 20),
    type: 'gradient',
  },
  {
    value: RED_BLUE_PALETTE.name,
    title: RED_BLUE_PALETTE.label,
    palette: colorPalette(RED_BLUE_PALETTE.colors, 20, true),
    type: 'gradient',
  },
  {
    value: YELLOW_ORANGE_RED_PALETTE.name,
    title: YELLOW_ORANGE_RED_PALETTE.label,
    palette: colorPalette(YELLOW_ORANGE_RED_PALETTE.colors, 20),
    type: 'gradient',
  },
  {
    value: YELLOW_GREEN_BLUE_PALETTE.name,
    title: YELLOW_GREEN_BLUE_PALETTE.label,
    palette: colorPalette(YELLOW_GREEN_BLUE_PALETTE.colors, 20),
    type: 'gradient',
  },
];
export const HEX_CONTRAST_COLOR = 0xffffff;
export const PIE_PALETTES = [
  {
    value: DEFAULT_PALETTE,
    title: 'Default',
    type: 'text',
  },
  {
    value: SINGLE_COLOR_PALETTE,
    title: 'Single Color',
    type: 'text',
  },
];

export const HEATMAP_PALETTE_COLOR = { name: REDS_PALETTE.label, color: REDS_PALETTE.label };
export const HEATMAP_SINGLE_COLOR = { name: 'singleColor', color: '#000000' };
export const OPACITY = 'opacity';
export const SPECTRUM = 'spectrum';
export const COLOR_BLACK = 'rgb(0,0,0)';
export const COLOR_WHITE = 'rgb(255,255,255)';
