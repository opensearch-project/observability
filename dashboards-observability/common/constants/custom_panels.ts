/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const CUSTOM_PANELS_API_PREFIX = '/api/observability/operational_panels';
export const CUSTOM_PANELS_DOCUMENTATION_URL = 'https://opensearch.org/docs/latest/observability-plugins/operational-panels/';
export const CREATE_PANEL_MESSAGE = 'Enter a name to describe the purpose of this custom panel.';
