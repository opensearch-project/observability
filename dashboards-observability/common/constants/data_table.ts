/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const GRID_HEADER_COLUMN_MAX_WIDTH = '150px';
export const GRID_PAGE_RANGE_DISPLAY = 5;
export const COLUMN_DEFAULT_MIN_WIDTH = 100;
export const GRID_PAGE_SIZES = [10, 50, 100];
export const ROW_DENSITIES = [
  { icon: 'tableDensityExpanded', height: 55, selected: false },
  { icon: 'tableDensityNormal', height: 45, selected: false },
  { icon: 'tableDensityCompact', height: 35, selected: true },
];

export const HEADER_HEIGHT = 35;
