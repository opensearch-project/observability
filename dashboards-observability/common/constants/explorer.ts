/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const EVENT_ANALYTICS_DOCUMENTATION_URL =
  'https://opensearch.org/docs/latest/observability-plugin/event-analytics/';
export const OPEN_TELEMETRY_LOG_CORRELATION_LINK =
  'https://opentelemetry.io/docs/reference/specification/logs/overview/#log-correlation';
export const RAW_QUERY = 'rawQuery';
export const FINAL_QUERY = 'finalQuery';
export const SELECTED_DATE_RANGE = 'selectedDateRange';
export const INDEX = 'index';
export const SELECTED_TIMESTAMP = 'selectedTimestamp';
export const SELECTED_FIELDS = 'selectedFields';
export const UNSELECTED_FIELDS = 'unselectedFields';
export const AVAILABLE_FIELDS = 'availableFields';
export const QUERIED_FIELDS = 'queriedFields';
export const TAB_ID_TXT_PFX = 'query-panel-';
export const TAB_TITLE = 'New query';
export const TAB_CHART_TITLE = 'Visualizations';
export const TAB_EVENT_TITLE = 'Events';
export const TAB_EVENT_ID_TXT_PFX = 'main-content-events-';
export const TAB_CHART_ID_TXT_PFX = 'main-content-vis-';
export const TAB_EVENT_ID = 'main-content-events';
export const TAB_CHART_ID = 'main-content-vis';
export const HAS_SAVED_TIMESTAMP = 'hasSavedTimestamp';
export const FILTER_OPTIONS = ['Visualization', 'Query'];
export const SAVED_QUERY = 'savedQuery';
export const SAVED_VISUALIZATION = 'savedVisualization';
export const SAVED_OBJECT_ID = 'savedObjectId';
export const SAVED_OBJECT_TYPE = 'objectType';
export const TAB_CREATED_TYPE = 'tabCreatedType';
export const NEW_TAB = 'newTab';
export const REDIRECT_TAB = 'redirect_tab';
export const PAGE_SIZE = 50;
export const DEFAULT_COLUMNS = ['', 'Time', '_source'];
export const OTEL_TRACE_ID = 'traceId';
export const DATE_PICKER_FORMAT = 'YYYY-MM-DD HH:mm:ss';
export const TIME_INTERVAL_OPTIONS = [
  {
    text: 'Minute',
    value: 'm',
  },
  {
    text: 'Hour',
    value: 'h',
  },
  {
    text: 'Day',
    value: 'd',
  },
  {
    text: 'Week',
    value: 'w',
  },
  {
    text: 'Month',
    value: 'M',
  },
  {
    text: 'Year',
    value: 'y',
  },
];

// redux
export const SELECTED_QUERY_TAB = 'selectedQueryTab';
export const QUERY_TAB_IDS = 'queryTabIds';
export const NEW_SELECTED_QUERY_TAB = 'newSelectedQueryTab';
export const REDUX_EXPL_SLICE_QUERIES = 'queries';
export const REDUX_EXPL_SLICE_QUERY_RESULT = 'queryResults';
export const REDUX_EXPL_SLICE_FIELDS = 'fields';
export const REDUX_EXPL_SLICE_QUERY_TABS = 'queryTabs';
export const REDUX_EXPL_SLICE_VISUALIZATION = 'explorerVisualization';
export const REDUX_EXPL_SLICE_COUNT_DISTRIBUTION = 'countDistributionVisualization';
export const PLOTLY_GAUGE_COLUMN_NUMBER = 5;
export const APP_ANALYTICS_TAB_ID_REGEX = /application-analytics-tab.+/;
export const DEFAULT_AVAILABILITY_QUERY = 'stats count() by span( timestamp, 1h )';
