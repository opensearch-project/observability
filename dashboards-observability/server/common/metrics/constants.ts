/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ComponentType, CounterType } from './types';

export const WINDOW = 3600;
export const INTERVAL = 60;
export const CAPACITY = (WINDOW / INTERVAL) * 2;
export const MILLIS_MULTIPLIER = 1000;

export const COMPONENTS = [
  'application_analytics',
  'operational_panels',
  'event_analytics',
  'notebooks',
  'trace_analytics',
  'metrics_analytics',
] as const;
export const REQUESTS = ['create', 'get', 'update', 'delete'] as const;

export const GLOBAL_BASIC_COUNTER: CounterType = (() => {
  const counter = {} as CounterType;
  COMPONENTS.forEach((component) => {
    counter[component] = {} as CounterType[ComponentType];
    REQUESTS.forEach((request) => {
      counter[component][request] = {
        total: 0,
      };
    });
  });
  return counter;
})();

export const DEFAULT_ROLLING_COUNTER: CounterType = (() => {
  const counter = {} as CounterType;
  COMPONENTS.forEach((component) => {
    counter[component] = {} as CounterType[ComponentType];
    REQUESTS.forEach((request) => {
      counter[component][request] = {
        count: 0,
        system_error: 0,
        user_error: 0,
      };
    });
  });
  return counter;
})();
