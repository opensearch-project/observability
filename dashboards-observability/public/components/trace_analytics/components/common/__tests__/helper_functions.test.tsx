/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { TEST_SERVICE_MAP, TEST_SERVICE_MAP_GRAPH } from '../../../../../../test/constants';
import {
  calculateTicks,
  filtersToDsl,
  fixedIntervalToMilli,
  fixedIntervalToTickFormat,
  getPercentileFilter,
  getServiceMapGraph,
  getServiceMapScaleColor,
  getServiceMapTargetResources,
  milliToNanoSec,
  minFixedInterval,
  MissingConfigurationMessage,
  nanoToMilliSec,
  NoMatchMessage,
  PanelTitle,
  renderBenchmark,
} from '../helper_functions';

describe('Helper functions', () => {
  configure({ adapter: new Adapter() });

  it('renders panel title', () => {
    const title = shallow(<PanelTitle title="test" totalItems={10} />);
    const titleZeroCount = shallow(<PanelTitle title="test" />);
    expect(title).toMatchSnapshot();
    expect(titleZeroCount).toMatchSnapshot();
  });

  it('renders no match and missing configuration messages', () => {
    const noMatchMessage = shallow(<NoMatchMessage size="s" />);
    const missingConfigurationMessage = shallow(<MissingConfigurationMessage />);
    expect(noMatchMessage).toMatchSnapshot();
    expect(missingConfigurationMessage).toMatchSnapshot();
  });

  it('renders benchmark', () => {
    // @ts-ignore
    const benchmarkPositive = mount(renderBenchmark(50));
    // @ts-ignore
    const benchmarkNegative = mount(renderBenchmark(-50));
    // @ts-ignore
    const benchmarkZero = mount(renderBenchmark(0));
    expect(benchmarkPositive).toMatchSnapshot();
    expect(benchmarkNegative).toMatchSnapshot();
    expect(benchmarkZero).toMatchSnapshot();
  });

  it('converts nanoseconds and milliseconds', () => {
    const ms = nanoToMilliSec(123456789);
    expect(ms).toEqual(123.456789);
    const ns = milliToNanoSec(123.456789);
    expect(ns).toEqual(123456789);
    // @ts-ignore
    const invalidMs = nanoToMilliSec('abc');
    expect(invalidMs).toEqual(0);
    // @ts-ignore
    const invalidNs = milliToNanoSec('abc');
    expect(invalidNs).toEqual(0);
  });

  it('returns service map scale color', () => {
    const color = getServiceMapScaleColor(0.5, 'latency');
    expect(color).toEqual('134, 105, 173');
  });

  it('returns service map graph', () => {
    const serviceMapGraph = getServiceMapGraph(TEST_SERVICE_MAP, 'latency', [
      0,
      50,
      100,
      150,
      200,
      250,
    ]);
    expect(serviceMapGraph).toEqual(TEST_SERVICE_MAP_GRAPH);
  });

  it('returns target resources by service name', () => {
    const targetResources = getServiceMapTargetResources(TEST_SERVICE_MAP, 'order');
    expect(targetResources).toEqual(['clear_order', 'update_order', 'get_order', 'pay_order']);
  });

  it('calculates ticks', () => {
    const ticks = calculateTicks(500, 200);
    const ticks2 = calculateTicks(0, 200, 10);
    expect(ticks).toEqual([0, 50, 100, 150, 200]);
    expect(ticks2).toEqual([0, 20, 40, 60, 80, 100, 120, 140, 160, 180, 200]);
  });

  it('calculates fixed_interval for date_histograms', () => {
    const fixedInterval = minFixedInterval('now-5y', 'now');
    expect(fixedInterval).toEqual('365d');
    const ms = fixedIntervalToMilli('1h');
    expect(ms).toEqual(3600000);
    const tickFormat = fixedIntervalToTickFormat('1h');
    expect(tickFormat).toEqual('');
  });

  it('returns percentile filter DSL', () => {
    const DSL = getPercentileFilter(
      [{ traceGroupName: 'order', durationFilter: { gte: 1000 } }],
      '>= 95th'
    );
    expect(DSL).toEqual(
      JSON.parse(
        `{"field":"Latency percentile within trace group","operator":"","value":">= 95th","inverted":false,"disabled":false,"custom":{"query":{"bool":{"must":[],"filter":[],"should":[{"bool":{"must":[{"term":{"traceGroup":{"value":"order"}}},{"range":{"traceGroupFields.durationInNanos":{"gte":1000}}}]}}],"must_not":[],"minimum_should_match":1}}}}`
      )
    );
  });

  it('converts filters to DSL', () => {
    const getTestDslFromFilters = (field = 'traceGroup', operator = 'exists') =>
      filtersToDsl(
        [
          {
            field,
            operator,
            value: { from: '100', to: '\u221E' },
            inverted: false,
            disabled: false,
          },
        ],
        'order',
        'now-5m',
        'now'
      );
    const existsDSL = getTestDslFromFilters();
    expect(JSON.stringify(existsDSL)).toEqual(
      '{"query":{"bool":{"must":[{"range":{"startTime":{"gte":"now-5m","lte":"now"}}},{"query_string":{"query":"order"}},{"exists":{"field":"traceGroup"}}],"filter":[],"should":[],"must_not":[]}},"custom":{"timeFilter":[{"range":{"startTime":{"gte":"now-5m","lte":"now"}}}],"serviceNames":[],"serviceNamesExclude":[],"traceGroup":[],"traceGroupExclude":[],"percentiles":{"query":{"bool":{"should":[]}}}}}'
    );

    const isDSL = getTestDslFromFilters('traceGroup', 'is');
    expect(JSON.stringify(isDSL)).toEqual(
      '{"query":{"bool":{"must":[{"range":{"startTime":{"gte":"now-5m","lte":"now"}}},{"query_string":{"query":"order"}},{"term":{"traceGroup":{"from":"100","to":"âˆž"}}}],"filter":[],"should":[],"must_not":[]}},"custom":{"timeFilter":[{"range":{"startTime":{"gte":"now-5m","lte":"now"}}}],"serviceNames":[],"serviceNamesExclude":[],"traceGroup":[],"traceGroupExclude":[],"percentiles":{"query":{"bool":{"should":[]}}}}}'
    );
    const isBetweenDSL = getTestDslFromFilters('durationInNanos', 'is between');
    expect(JSON.stringify(isBetweenDSL)).toEqual(
      `{"query":{"bool":{"must":[{"range":{"startTime":{"gte":"now-5m","lte":"now"}}},{"query_string":{"query":"order"}},{"range":{"durationInNanos":{"gte":"100"}}}],"filter":[],"should":[],"must_not":[]}},"custom":{"timeFilter":[{"range":{"startTime":{"gte":"now-5m","lte":"now"}}}],"serviceNames":[],"serviceNamesExclude":[],"traceGroup":[],"traceGroupExclude":[],"percentiles":{"query":{"bool":{"should":[]}}}}}`
    );

    const customDSL = filtersToDsl(
      [
        {
          field: 'serviceName',
          operator: 'is',
          value: 'order',
          inverted: false,
          disabled: false,
          custom: { query: { bool: { should: ['test'], minimum_should_match: 1 } } },
        },
      ],
      'order',
      'now-5m',
      'now'
    );
    expect(JSON.stringify(customDSL)).toEqual(
      `{"query":{"bool":{"must":[{"range":{"startTime":{"gte":"now-5m","lte":"now"}}},{"query_string":{"query":"order"}}],"filter":[],"should":["test"],"must_not":[],"minimum_should_match":1}},"custom":{"timeFilter":[{"range":{"startTime":{"gte":"now-5m","lte":"now"}}}],"serviceNames":[],"serviceNamesExclude":[],"traceGroup":[],"traceGroupExclude":[],"percentiles":{"query":{"bool":{"should":["test"],"minimum_should_match":1}}}}}`
    );
  });
});
