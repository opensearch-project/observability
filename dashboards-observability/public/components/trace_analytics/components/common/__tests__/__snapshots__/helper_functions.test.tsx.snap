// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper functions renders benchmark 1`] = `
<EuiText
  size="s"
  style={
    Object {
      "color": "#c23f25",
    }
  }
>
  <div
    className="euiText euiText--small"
    style={
      Object {
        "color": "#c23f25",
      }
    }
  >
    50% ▴
  </div>
</EuiText>
`;

exports[`Helper functions renders benchmark 2`] = `
<EuiText
  size="s"
  style={
    Object {
      "color": "#3f7e23",
    }
  }
>
  <div
    className="euiText euiText--small"
    style={
      Object {
        "color": "#3f7e23",
      }
    }
  >
    50% ▾
  </div>
</EuiText>
`;

exports[`Helper functions renders benchmark 3`] = `
<EuiText
  size="s"
  style={
    Object {
      "color": "#9ea8a9",
    }
  }
>
  <div
    className="euiText euiText--small"
    style={
      Object {
        "color": "#9ea8a9",
      }
    }
  >
    0% -
  </div>
</EuiText>
`;

exports[`Helper functions renders no match and missing configuration messages 1`] = `
<Fragment>
  <EuiSpacer
    size="s"
  />
  <EuiEmptyPrompt
    body={
      <EuiText>
        No data matches the selected filter. Clear the filter and/or increase the time range to see more results.
      </EuiText>
    }
    title={
      <h2>
        No matches
      </h2>
    }
  />
  <EuiSpacer
    size="s"
  />
</Fragment>
`;

exports[`Helper functions renders no match and missing configuration messages 2`] = `
<Fragment>
  <EuiEmptyPrompt
    actions={
      <EuiButton
        color="primary"
        iconSide="right"
        iconType="popout"
        onClick={[Function]}
      >
        Learn more
      </EuiButton>
    }
    body={
      <EuiText>
        The indices required for trace analytics (otel-v1-apm-span-* and otel-v1-apm-service-map*) do not exist or you do not have permission to access them.
      </EuiText>
    }
    title={
      <h2>
        Trace Analytics not set up
      </h2>
    }
  />
</Fragment>
`;

exports[`Helper functions renders panel title 1`] = `
<EuiText
  size="m"
>
  <span
    className="panel-title"
  >
    test
  </span>
  <span
    className="panel-title-count"
  >
     (10)
  </span>
</EuiText>
`;

exports[`Helper functions renders panel title 2`] = `
<EuiText
  size="m"
>
  <span
    className="panel-title"
  >
    test
  </span>
</EuiText>
`;
